### PR TITLE
PDP Revisions and Revised

### LINKED ISSUE #1178: Product detail page editing workflow improvements
Simple publishing workflow for products and variants.

Goals:
- [ ] Allow the admin users to edit a product without having changes showing up immediately for all users.
- [ ] Save any changes to a product as a draft revision in the `Revisions` Collection.
- [ ] Use a "Publish Changes" button to publish all draft changes to a product and all variants with a draft.

### PR DESCRIPTION
**Pull request by mikemurray**
Resolves #1178 

Major points
- Add revisions to product detail page
- React version of product detail page
- React version of product admin view
- Revision control can be enabled / disabled through settings

### CODE REVIEW

**Comment by aaronjudd** [comment]
console error:

```
warning.js:36 Warning: Invalid argument supplied to oneOf, expected an instance of array.
```

**Comment by aaronjudd** [comment]
Both pages (old and new) are still present... updating coming?

File: .meteor/versions [hunk 1]
```diff
@@ -78,6 +78,7 @@ htmljs@1.0.10
 http@1.2.9
 id-map@1.0.8
 jeremy:stripe@1.6.0
+johanbrook:publication-collector@1.0.0
```
Comment (aaronjudd) [inline .meteor/versions]: this is required for testing?

File: .reaction/config.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (aaronjudd) [inline .reaction/config.json]: @jshimko  `.reaction/config.json`  supposed to be committed?

File: imports/plugins/core/revisions/server/startup.js [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+
+    modifier.$set = newSet;
+    modifier.$inc = newInc;
+  }
+
+  // prevent the underlying document from being modified as it is in draft mode
+  return false;
+});
+
+// Products.after.update(function (userId, product, fieldNames, modifier, options) {
```
Comment (aaronjudd) [inline imports/plugins/core/revisions/server/startup.js]: Not used? delete `Products.after.update`?

File: imports/plugins/core/ui/client/components/button/button.jsx [hunk 4]
```diff
@@ -32,39 +59,68 @@ class Button extends Component {
...
+    }
+
+    return null;
+  }
+
   render() {
     const classes = classnames({
       "btn": true,
-      "btn-default": this.props.status === null || this.props.status === "default",
+      "btn-default": this.props.status === null || this.props.status === undefined || this.props.status === "default",
```
Comment (aaronjudd) [inline imports/plugins/core/ui/client/components/button/button.jsx]: should this check be `!this.props.status` or `=== "undefined"`

File: imports/plugins/core/ui/client/providers/translationProvider.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  children: PropTypes.node,
+  translations: PropTypes.object.isRequired
+};
+
+function composer(props, onData) {
+  i18nextDep.depend();
+
+  onData(null, {
+    translations: {
+      language: Session.get("language")
```
Comment (aaronjudd) [inline imports/plugins/core/ui/client/providers/translationProvider.js]: Session.get("language") is deprecated in #1338

File: .reaction/config.json (hunk 2 above)
Comment (jshimko) [inline .reaction/config.json]: Nope, it was gitignored (likely during the time work on this branch was happening)

**Comment by mikemurray** [comment]
@aaronjudd yes, old and new are still in, ready to remove the duplicate once I'm sure they both work the same.

**Comment by aaronjudd** [comment]
Needs `meteor npm install --save react-dnd react-dropzone`

**Comment by aaronjudd** [comment]
When attempting to add a new product tag.
``` 
Exception while invoking method 'products/updateProductTags' Error: Match error: Expected string, got null
```

**Comment by swrdfish** [comment]
Got the following error while trying out this branch:

```
Unable to resolve some modules:

  "lodash/camelcase" in
/home/binayak/projects/reaction/imports/plugins/core/ui/client/components/translation/translation.js
(web.browser)
```

The issue is not solved with:
` meteor npm install --save lodash react-dnd react-dropzone`

**Comment by swrdfish** [comment]
Got the following error while trying out this branch:
```
Unable to resolve some modules:

  "lodash/camelcase" in
/home/binayak/projects/reaction/imports/plugins/core/ui/client/components/translation/translation.js
(web.browser)
```

The following didn't solve the issue:
`  meteor npm install --save lodash react-dnd react-dropzone `

**Comment by aaronjudd** [comment]
@swrdfish you might need to do `npm i --save lodash.camelcase` 
@mikemurray can you please update (and commit) the package.json with any additional dependencies that you have?

**Comment by swrdfish** [comment]
@aaronjudd @mikemurray  tried that, still getting the same error. Any other ideas?

**Comment by mikemurray** [comment]
It seems `icon` folder and files are being ignored because of some .`gitignore` rule; though, there is no such rule in any of our `.gitignore` files. Could just be my system.

I've added them by force for the moment.

**Comment by aaronjudd** [comment]
First image added is fine, without error, however on add of an additional image on the variant, but only when using the side controls and clicking "Drop file to Upload":

```
I20160926-12:47:52.653(-7)? Exception while invoking method 'products/updateProductField' Error: After filtering out keys not in the schema, your modifier is now empty
I20160926-12:47:52.653(-7)?     at [object Object].doValidate (packages/aldeed_collection2-core/lib/collection2.js:387:1)
I20160926-12:47:52.653(-7)?     at [object Object].Mongo.Collection.(anonymous function) [as update] (packages/aldeed_collection2-core/lib/collection2.js:185:1)
I20160926-12:47:52.654(-7)?     at [object Object].productsUpdateProductField (server/methods/catalog.js:801:29)
I20160926-12:47:52.654(-7)?     at packages/check/match.js:107:1
I20160926-12:47:52.654(-7)?     at [object Object]._.extend.withValue (packages/meteor/dynamics_nodejs.js:56:1)
I20160926-12:47:52.655(-7)?     at Object.exports.Match._failIfArgumentsAreNotAllChecked (packages/check/match.js:106:1)
I20160926-12:47:52.655(-7)?     at maybeAuditArgumentChecks (packages/ddp-server/livedata_server.js:1708:18)
I20160926-12:47:52.655(-7)?     at packages/ddp-server/livedata_server.js:711:19
I20160926-12:47:52.655(-7)?     at [object Object]._.extend.withValue (packages/meteor/dynamics_nodejs.js:56:1)
I20160926-12:47:52.656(-7)?     at packages/ddp-server/livedata_server.js:709:40
I20160926-12:47:52.656(-7)?     at [object Object]._.extend.withValue (packages/meteor/dynamics_nodejs.js:56:1)
I20160926-12:47:52.656(-7)?     at packages/ddp-server/livedata_server.js:707:46
I20160926-12:47:52.657(-7)?     at Session.method (packages/ddp-server/livedata_server.js:681:23)
I20160926-12:47:52.657(-7)?     at packages/ddp-server/livedata_server.js:551:43
```

**Comment by aaronjudd** [comment]
Typo
screenshot 2016-09-26 12 57 26

**Comment by aaronjudd** [comment]
Not sure I understand what status the "eye" is meant to convey.. it seems to just toggle back and forth whenever anything changes...

2vxngmurau

**Comment by aaronjudd** [comment]
Deleting a variant produced a strange effect, wasn't sure if it was because the "deletion" has not been published and if there is some workflow around that that just isn't clear, but the image is deleted regardless.. but not the variant from this view.

2vxngmurau

**Comment by aaronjudd** [comment]
need some product brainstorming around this..
basic_reaction_product

**Comment by aaronjudd** [comment]
A few items on the social shares:
## facebook
`Uncaught ReferenceError: FB is not defined`

## pinterest
pinterest_ _the_world s_catalog_of_ideas

## g+
`The requested URL was not found on this server. That’s all we know.`

**Comment by aaronjudd** [comment]
Weird toggle behaviour when attempting to select and edit child options..

**Comment by mikemurray** [comment]
Summary of changes today
- Fix for broken pinterest image
- Fix for variant selection / edit issues
- Fix for undefined `FB`
- Added a badge for deleted products to make it clearer that they are going to be deleted once changes are published.
- Fix for broken product subscription, catalog, and inventory tests due to revision control on / off state.

**Comment by aaronjudd** [comment]
Typos in `  "click .js-child-varaint-heading": function () { ` let's fix and is likely to cause issue (like it that really the class?)

**Comment by aaronjudd** [comment]
The image upload when using the "drop file to upload" is caused by form input triggered on `  "change .child-variant-form :input":`.  Let's get more specific there and exclude that event..

**Comment by hrath2015** [comment]
Noticed this also. To Me it looks like typo. ``.php.header`` should be ``.pdp.header``
https://github.com/reactioncommerce/reaction/blob/pdp-workflow-1178/imports/plugins/included/default-theme/client/styles/products/productDetail.less#L19

**Comment by aaronjudd** [comment]
Blocker for merge to development:
- variant details not updating on change (of a parent variant)

Let's create issues for follow up tasks:

- [ ] Create Issue for Image Version (media)
- [ ] Create Issue for Tag versioning
- [ ] Create issue for publish workflow UX
- [ ] move revisions card into Catalog card  (and i18n update)
- [ ] update "remove variant" confirm to SWAL2
- [ ] updated "Loading" spinner
- [ ] UX move delete button to same row (shouldn't need row to be active to delete)
- [ ] UX to collapse accordions, focus edit target and open

**Comment by mikemurray** [comment]
thanks @hrath2015

**Comment by aaronjudd** [comment]
👍  approving for merge to development.  I'll create issues for the remaining tasks as noted

**Comment by kieckhafer** [comment]
👍 looks good to merge to Dev and test there

**Comment by hrath2015** [comment]
👍 Great work with revamped PDP.  Is there **any design doc or functional chart/doc available which can be referred  to do  testing around it.**  Some findings: 1.) Image stores are not referred 2.) placeholders are missing.  How to go back to a previous revision. Once  product/variant is published is it allowed and some more questions..

**Comment by aaronjudd** [comment]
@hrath2015 re: remaining items -> I'm going to merge to development and we'll make issues for the items I noted, image revisions is one of those item.  I think we can just store some revision status in the image metadata to implement revisioning with images.  Not sure what you mean on the placeholders are missing -> in the inputs?  also missing is the former visual que of turning the field green on successful input.   We're going to be jamming on the UX of this next week, while functional it's a bit confusing, and I think we can address these in a focused UX session...

**Comment by hrath2015** [comment]
@aaronjudd thanks a lot for updates. I see a plugin Product Admin(in progress) for product specific admin activities.    Will continue to test from dev branch and report findings.

**Pull request merged**